src/models/highlightedtext.cpp

```diff
@@ -0,0 +1,200 @@
+/*
+    SPDX-FileCopyrightText: Lieven Hey <lieven.hey@kdab.com>
+    SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#include "highlightedtext.h"
+
+#include <QPalette>
+
+#include <memory>
+
+#include <KColorScheme>
+#if KFSyntaxHighlighting_FOUND
+#include <KSyntaxHighlighting/AbstractHighlighter>
+#include <KSyntaxHighlighting/Definition>
+#include <KSyntaxHighlighting/Format>
+#include <KSyntaxHighlighting/Repository>
+#include <KSyntaxHighlighting/State>
+#include <KSyntaxHighlighting/SyntaxHighlighter>
+#include <KSyntaxHighlighting/Theme>
+#endif
+
+#include "formattingutils.h"
+
+namespace {
+#if KFSyntaxHighlighting_FOUND
+class KSyntaxHighlightingImplementation : public HighlightingImplementation,
+                                          public KSyntaxHighlighting::AbstractHighlighter
+{
+public:
+    KSyntaxHighlightingImplementation(KSyntaxHighlighting::Repository* repository)
+        : m_repository(repository)
+    {
+    }
+    ~KSyntaxHighlightingImplementation() override = default;
+
+    QVector<QTextLayout::FormatRange> format(const QStringList& text) override
+    {
+        m_formats.clear();
+        m_offset = 0;
+
+        KSyntaxHighlighting::State state;
+        for (const auto& line : text) {
+            state = highlightLine(line, state);
+
+            // KSyntaxHighlighting uses line offsets but QTextLayout uses global offsets
+            m_offset += line.size();
+        }
+
+        return m_formats;
+    }
+
+    void themeChanged() override
+    {
+        if (!m_repository) {
+            return;
+        }
+
+        KSyntaxHighlighting::Repository::DefaultTheme theme;
+        if (QPalette().base().color().lightness() < 128) {
```

Comment:
this cannot work, can it? see: https://doc.qt.io/qt-6/qpalette.html#QPalette

you want to use https://doc.qt.io/qt-6/qguiapplication.html#palette or something instead, no?

Reply:
good to know, I copied that code from another location, will fix both